ambry-router/src/test/java/com.github.ambry.router/DefaultGCMCryptoServiceTest.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.utils.TestUtils;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.util.Arrays;
+import java.util.Properties;
+import javax.crypto.spec.SecretKeySpec;
+import org.bouncycastle.util.encoders.Hex;
+import org.junit.Assert;
+import org.junit.Test;
+
+import static com.github.ambry.router.CryptoTestUtils.*;
+
+
+/**
+ * Tests {@link DefaultGCMCryptoService} and {@link DefaultCryptoServiceFactory}
+ */
+public class DefaultGCMCryptoServiceTest {
+
+  private static final int MAX_DATA_SIZE = 10000;
+  private static final int DEFAULT_KEY_SIZE_IN_CHARS = 64;
+
+  /**
+   * Tests basic encryption and decryption for different sizes of keys and random data in bytes
+   */
+  @Test
+  public void testEncryptDecryptBytes() throws Exception {
+    for (int j = 0; j < 3; j++) {
+      String key = getRandomKey(DEFAULT_KEY_SIZE_IN_CHARS);
+      Properties props = getKMSProperties(key, DEFAULT_KEY_SIZE_IN_CHARS);
+      VerifiableProperties verifiableProperties = new VerifiableProperties((props));
+      SecretKeySpec secretKeySpec = new SecretKeySpec(Hex.decode(key), "AES");
+      CryptoService<SecretKeySpec> cryptoService =
+          new DefaultCryptoServiceFactory(verifiableProperties).getCryptoService();
+      for (int i = 0; i < 5; i++) {
+        int size = TestUtils.RANDOM.nextInt(MAX_DATA_SIZE);
+        byte[] randomData = new byte[size];
+        TestUtils.RANDOM.nextBytes(randomData);
+        ByteBuffer toEncrypt = ByteBuffer.wrap(randomData);
+        ByteBuffer encryptedBytes = cryptoService.encrypt(toEncrypt, secretKeySpec);
+        Assert.assertFalse("Encrypted bytes and plain bytes should not match",
```

Comment:
wouldn't checking the size be enough? On those lines, it is not verified that the data is not the same, say after a few leading bytes. 
What I am getting at is that, is a full array check really necessary? Are there any reliable ways of ensuring that the content is _truly_ encrypted?

Reply:
Also it looks like similar checks don't exist for the key test

Reply:
@vgkholla : Nope I couldn't find of a way to ensure that the content is encrypted apart from checking that the content differs from original data. Please let me know if you can think of something. But if we know how much is the size of iv, isn't that leakage of info. That is internal to the impl and hence the tests shouldn't really use that info.